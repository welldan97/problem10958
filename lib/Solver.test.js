const Solver = require('./Solver');
const { allDigitsByOneGenerator } = require('../specifics/numberGenerators');
const { add, multiply } = require('../specifics/operations');

const present = value => ({
  ...value,
  operations: value.operations.map(o => o.name),
});

describe('Solver', () => {
  it('finds all combinations', () => {
    const numberGenerator = allDigitsByOneGenerator(3);
    const operations = [add, multiply];
    const values = [];
    Solver.run(numberGenerator, operations, value => {
      values.push(present(value));
      return false;
    });
    expect(values).toMatchSnapshot();
  });

  it('finishes once condition met', () => {
    const numberGenerator = allDigitsByOneGenerator(3);
    const operations = [add, multiply];
    const values = [];
    Solver.run(numberGenerator, operations, value => {
      values.push(present(value));
      if (value.result === 6) return true;
    });
    expect(values).toMatchSnapshot();
  });

  it('finds soltions for given batch', () => {
    const numberGenerator = allDigitsByOneGenerator(3);
    const operations = [add, multiply];
    const values = [];
    Solver.run(
      numberGenerator,
      operations,
      value => {
        values.push(present(value));
        return false;
      },
      { batchLength: 3, batchNumber: 1 },
    );
    expect(values).toMatchSnapshot();
  });
});
