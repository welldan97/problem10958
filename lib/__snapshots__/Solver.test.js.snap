// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Solver finds all combinations 1`] = `
Array [
  Object {
    "index": 0,
    "numbers": Uint8Array [
      1,
    ],
    "operations": Array [],
    "priorities": Array [],
    "result": 1,
  },
  Object {
    "index": 1,
    "numbers": Uint8Array [
      1,
      2,
    ],
    "operations": Uint8Array [
      0,
    ],
    "priorities": Uint8Array [
      0,
    ],
    "result": 3,
  },
  Object {
    "index": 2,
    "numbers": Uint8Array [
      1,
      2,
    ],
    "operations": Uint8Array [
      1,
    ],
    "priorities": Uint8Array [
      0,
    ],
    "result": 2,
  },
  Object {
    "index": 3,
    "numbers": Uint8Array [
      1,
      2,
      3,
    ],
    "operations": Uint8Array [
      0,
      0,
    ],
    "priorities": Uint8Array [
      0,
      1,
    ],
    "result": 6,
  },
  Object {
    "index": 4,
    "numbers": Uint8Array [
      1,
      2,
      3,
    ],
    "operations": Uint8Array [
      0,
      0,
    ],
    "priorities": Uint8Array [
      1,
      0,
    ],
    "result": 6,
  },
  Object {
    "index": 5,
    "numbers": Uint8Array [
      1,
      2,
      3,
    ],
    "operations": Uint8Array [
      0,
      1,
    ],
    "priorities": Uint8Array [
      0,
      1,
    ],
    "result": 9,
  },
  Object {
    "index": 6,
    "numbers": Uint8Array [
      1,
      2,
      3,
    ],
    "operations": Uint8Array [
      0,
      1,
    ],
    "priorities": Uint8Array [
      1,
      0,
    ],
    "result": 7,
  },
  Object {
    "index": 7,
    "numbers": Uint8Array [
      1,
      2,
      3,
    ],
    "operations": Uint8Array [
      1,
      0,
    ],
    "priorities": Uint8Array [
      0,
      1,
    ],
    "result": 5,
  },
  Object {
    "index": 8,
    "numbers": Uint8Array [
      1,
      2,
      3,
    ],
    "operations": Uint8Array [
      1,
      0,
    ],
    "priorities": Uint8Array [
      1,
      0,
    ],
    "result": 5,
  },
  Object {
    "index": 9,
    "numbers": Uint8Array [
      1,
      2,
      3,
    ],
    "operations": Uint8Array [
      1,
      1,
    ],
    "priorities": Uint8Array [
      0,
      1,
    ],
    "result": 6,
  },
  Object {
    "index": 10,
    "numbers": Uint8Array [
      1,
      2,
      3,
    ],
    "operations": Uint8Array [
      1,
      1,
    ],
    "priorities": Uint8Array [
      1,
      0,
    ],
    "result": 6,
  },
]
`;

exports[`Solver finds soltions for given batch 1`] = `
Array [
  Object {
    "index": 3,
    "numbers": Uint8Array [
      1,
      2,
      3,
    ],
    "operations": Uint8Array [
      0,
      0,
    ],
    "priorities": Uint8Array [
      0,
      1,
    ],
    "result": 6,
  },
  Object {
    "index": 4,
    "numbers": Uint8Array [
      1,
      2,
      3,
    ],
    "operations": Uint8Array [
      0,
      0,
    ],
    "priorities": Uint8Array [
      1,
      0,
    ],
    "result": 6,
  },
  Object {
    "index": 5,
    "numbers": Uint8Array [
      1,
      2,
      3,
    ],
    "operations": Uint8Array [
      0,
      1,
    ],
    "priorities": Uint8Array [
      0,
      1,
    ],
    "result": 9,
  },
]
`;

exports[`Solver finishes once condition met 1`] = `
Array [
  Object {
    "index": 0,
    "numbers": Uint8Array [
      1,
    ],
    "operations": Array [],
    "priorities": Array [],
    "result": 1,
  },
  Object {
    "index": 1,
    "numbers": Uint8Array [
      1,
      2,
    ],
    "operations": Uint8Array [
      0,
    ],
    "priorities": Uint8Array [
      0,
    ],
    "result": 3,
  },
  Object {
    "index": 2,
    "numbers": Uint8Array [
      1,
      2,
    ],
    "operations": Uint8Array [
      1,
    ],
    "priorities": Uint8Array [
      0,
    ],
    "result": 2,
  },
  Object {
    "index": 3,
    "numbers": Uint8Array [
      1,
      2,
      3,
    ],
    "operations": Uint8Array [
      0,
      0,
    ],
    "priorities": Uint8Array [
      0,
      1,
    ],
    "result": 6,
  },
]
`;
